# Tugas

- Lakukan klasifikasi pada data MNIST dengan menggunakan model ANN
- Anda diperbolehkan melakukan eksplorasi terhadap,
  - Metode pra pengolahan
  - Pemilihan fitur
  - Arsitektur ANN
  - Fungsi Aktiviasi
- ANN diimplementasikan dengan menggunakan tensorflow.
- DIKERJAKAN SECARA BERKELOMPOK
- JELASKAN HASIL YANG ANDA DAPATKAN,
  - AKURASI
  - CONFUSION MATRIX
  - KONFIGURASI MODEL --> MULAI DARI PRA PENGOLAHAN SAMPAI ARSITEKTUR ANNN

In [2]:
import tensorflow as tf
from keras import layers, models
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix

c:\Users\Hp\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [3]:
# Mengunduh dataset MNIST
mnist = fetch_openml('mnist_784')

# Memisahkan data gambar dan label
images = mnist.data.astype('float32')
labels = mnist.target.astype('int')

# Skalakan data
images /= 255.0 

# Membagi data menjadi data pelatihan dan pengujian
train_images, test_images, train_labels, test_labels = train_test_split(
    images, labels, test_size=0.2, random_state=42)

# Mendefinisikan arsitektur model ANN
model = models.Sequential([
    layers.Reshape(target_shape=(28, 28, 1), input_shape=(784,)),
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

# Melakukan kompilasi terhadap model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Melatih model
history = model.fit(train_images, train_labels,
                    epochs=10, validation_split=0.2)

# Melakukan evaluasi model pada data pengujian
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Akurasi pada data pengujian: {test_acc:.2f}')

c:\Users\Hp\anaconda3\lib\site-packages\pandas\core\arrays\categorical.py:528: RuntimeWarning: invalid value encountered in cast
  fill_value = lib.item_from_zerodim(np.array(np.nan).astype(dtype))


Epoch 1/10
1400/1400 [==============================] - 32s 22ms/step - loss: 0.1842 - accuracy: 0.9436 - val_loss: 0.0804 - val_accuracy: 0.9740
Epoch 2/10
1400/1400 [==============================] - 31s 22ms/step - loss: 0.0584 - accuracy: 0.9819 - val_loss: 0.0520 - val_accuracy: 0.9826
Epoch 3/10
1400/1400 [==============================] - 39s 28ms/step - loss: 0.0359 - accuracy: 0.9889 - val_loss: 0.0575 - val_accuracy: 0.9807
Epoch 4/10
1400/1400 [==============================] - 33s 23ms/step - loss: 0.0261 - accuracy: 0.9918 - val_loss: 0.0531 - val_accuracy: 0.9837
Epoch 5/10
1400/1400 [==============================] - 32s 23ms/step - loss: 0.0168 - accuracy: 0.9943 - val_loss: 0.0552 - val_accuracy: 0.9846
Epoch 6/10
1400/1400 [==============================] - 32s 23ms/step - loss: 0.0126 - accuracy: 0.9958 - val_loss: 0.0638 - val_accuracy: 0.9839
Epoch 7/10
1400/1400 [==============================] - 33s 23ms/step - loss: 0.0124 - accuracy: 0.9960 - val_loss: 0.0650 -

In [3]:
# Melakukan evaluasi terhadap akurasi data pelatihan
train_loss, train_acc = model.evaluate(train_images, train_labels)
print(f'Akurasi pada data pelatihan: {train_acc:.2f}')

1750/1750 [==============================] - 9s 5ms/step - loss: 0.0140 - accuracy: 0.9968
Akurasi pada data pelatihan: 1.00


In [4]:
# Membuat prediksi pada data pengujian
predictions = model.predict(test_images)
predicted_labels = np.argmax(predictions, axis=1)

# Membuat matriks kebingungan
confusion = confusion_matrix(test_labels, predicted_labels)
print('Matriks Kebingungan:')
print(confusion)

438/438 [==============================] - 2s 5ms/step
Matriks Kebingungan:
[[1327    1    3    0    0    2    6    2    2    0]
 [   0 1592    3    0    1    0    0    3    1    0]
 [   0    5 1363    3    2    0    0    4    1    2]
 [   0    3    7 1408    0    6    0    4    1    4]
 [   1    0    1    0 1277    0    0    3    0   13]
 [   1    0    0    6    0 1255    7    0    4    0]
 [   4    0    0    0    3    6 1382    0    0    1]
 [   1    2    7    0    3    0    0 1487    0    3]
 [   0    5   11    2    3    5    4    5 1310   12]
 [   5    1    1    4    7    4    0   10    2 1386]]


### **KESIMPULAN AKHIR**

1. Akurasi

  - Hasil akurasi : 0.98
  - Menunjukkan akurasi yang tinggi dan dapat dikatakan sebagai
    prediksi data yang baik

2. Confusion Matrix

  - Matriks yang memiliki ukuran 10x10 sel, yaitu 10 barisa dan 10 kolom
  - Pada indeks ke-0 atau Baris pertama memberi gambaran pada kelas 0 yang sebenarnya, di dalamnya terdapat 1329 sampel asli dan beberapa yang salah diklasifikasikan sebagai kelas lain. begitu pula untuk 9 baris berikutnya.
  - Pada kolom pertama (indeks 0) memberi gambaran terhadap prediksi kelas 0 yag berisikan 1338 prediksi benar, dan beberapa yang salah. Begitu pula pada kolom-kolom berikutnya hingga kelas 9

3. Konfigurasi Model
  - Pra-pemrosesan
        
    - mengubah rentang nilai piksel  gambar.
    - Membagi data gambar dan label ke dalam dua set terpisah (data pelatihan dan data pengujian)
  
  - Mendifinisikan Arsitektur Model ANN

    - Mendefinisikan arsitektur dari model ANN, meliputi lapisan konvolusi, max pooling, dan fully connected, yang akan digunakan untuk melakukan klasifikasi gambar.
  
  - Kompilasi Model

    -  Melakukan kompilasi terhadap model neural network dan menentukan bagaimana model akan melakukan pembelajaran dan evaluasi selama proses pelatihan.

  - Melatih Model

    - Pada tahapan ini hasil pelatihan akan disimpan dalam objek history, yang digunakan untuk menganalisis dan memvisualisasikan kinerja model selama pelatihan. Ketika tahap pelatihan telah selesai, model akan siap digunakan untuk menemukan prediksi pada data.

  - Evaluasi Model pada Data Pengujian

    - Menampilkan hasil seberapa baik model yang dijalankan dalam melakukan klasifikasi pada data baru dan tidak terlihat sebelumnya.